File: backend/app/routers/notifications.py
import uuid
from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, update, func

from app.database import get_db
from app.models.notification import Notification
from app.utils.auth import get_current_business_id

router = APIRouter()


@router.get("/")
async def list_notifications(
    limit: int = 20,
    unread_only: bool = False,
    business_id: str = Depends(get_current_business_id),
    db: AsyncSession = Depends(get_db),
):
    bid = uuid.UUID(business_id)
    query = select(Notification).where(Notification.business_id == bid)
    if unread_only:
        query = query.where(Notification.is_read == False)  # noqa: E712
    query = query.order_by(Notification.created_at.desc()).limit(limit)
    result = await db.execute(query)
    notifications = result.scalars().all()
    return [
        {
            "id": str(n.id),
            "title": n.title,
            "message": n.message,
            "category": n.category,
            "is_read": n.is_read,
            "link": n.link,
            "created_at": n.created_at.isoformat(),
        }
        for n in notifications
    ]


@router.get("/unread-count")
async def unread_count(
    business_id: str = Depends(get_current_business_id),
    db: AsyncSession = Depends(get_db),
):
    bid = uuid.UUID(business_id)
    result = await db.execute(
        select(func.count()).where(
            Notification.business_id == bid, Notification.is_read == False  # noqa: E712
        )
    )
    return {"count": result.scalar()}


@router.patch("/{notification_id}/read")
async def mark_read(
    notification_id: uuid.UUID,
    business_id: str = Depends(get_current_business_id),
    db: AsyncSession = Depends(get_db),
):
    await db.execute(
        update(Notification)
        .where(Notification.id == notification_id, Notification.business_id == uuid.UUID(business_id))
        .values(is_read=True)
    )
    await db.commit()
    return {"status": "ok"}


@router.post("/mark-all-read")
async def mark_all_read(
    business_id: str = Depends(get_current_business_id),
    db: AsyncSession = Depends(get_db),
):
    await db.execute(
        update(Notification)
        .where(Notification.business_id == uuid.UUID(business_id), Notification.is_read == False)  # noqa: E712
        .values(is_read=True)
    )
    await db.commit()
    return {"status": "ok"}

File: frontend/src/lib/stores/notifications.ts
import { writable, get } from 'svelte/store';
import { api } from '$lib/api';
import { auth } from './auth';

export interface Notification {
	id: string;
	title: string;
	message: string;
	category: string;
	is_read: boolean;
	link: string | null;
	created_at: string;
}

export const notifications = writable<Notification[]>([]);
export const unreadCount = writable(0);

export async function fetchNotifications() {
	const { token } = get(auth);
	if (!token) return;
	try {
		const data = await api<Notification[]>('/notifications/?limit=20', { token });
		notifications.set(data);
	} catch {
		// API not ready
	}
}

export async function fetchUnreadCount() {
	const { token } = get(auth);
	if (!token) return;
	try {
		const data = await api<{ count: number }>('/notifications/unread-count', { token });
		unreadCount.set(data.count);
	} catch {
		// API not ready
	}
}

export async function markRead(notificationId: string) {
	const { token } = get(auth);
	if (!token) return;
	await api(`/notifications/${notificationId}/read`, { method: 'PATCH', token });
	notifications.update((items) =>
		items.map((n) => (n.id === notificationId ? { ...n, is_read: true } : n))
	);
	unreadCount.update((c) => Math.max(0, c - 1));
}

export async function markAllRead() {
	const { token } = get(auth);
	if (!token) return;
	await api('/notifications/mark-all-read', { method: 'POST', token });
	notifications.update((items) => items.map((n) => ({ ...n, is_read: true })));
	unreadCount.set(0);
}
